WordPress/Classes/ViewRelated/Aztec/Processors/CalypsoProcessorIn.swift

```diff
@@ -13,6 +13,13 @@ class CalypsoProcessorIn: Processor {
     /// Current as of 2017/08/08
     ///
     func process(_ text: String) -> String {
+
+        // This is awful, but our APIs don't offer any flag to know if the content is Gutenberg content or not.
```

Comment:
I agree, this is awful. Did we make a request to the Gutenberg team to add a flag to Gutenberg posts? @pento @mcsf I'm sure that would come in handy for other platforms too, for example when opening a Gutenberg post in Calypso.

Reply:
> this is awful

Just so I can better answer: which part are you referring to? The heuristic, or the skipping of the routine?

Reply:
> Just so I can better answer: which part are you referring to? The heuristic, or the skipping of the routine?

It's awful because:

- **Redundancy & syncrhonization**: We have to implement this logic at the client level.  External consumers of our APIs (such as WPiOS and WP Android) must duplicate this logic, or implement their own version of it which results in an inherent problem of synchronization (such as Aztec stripping `p` tags when it should not).
- **Possible data loss due to an imperfect test condition**: The `wp_content` field must be treated differently depending on the type of content, since for older posts we must automatically add and remove `p` tags, but for Gutenberg posts we need to leave them unchanged.  The current check is not perfect and if a post written in `wp-admin` describing Gutenberg included "<-- wp:" in its content, it would cause issues and probably data loss.
- If we ever change other things in terms of `wp_content` formatting within Gutenberg, the lack of explicit post type / versioning will make it impossible for us to warn users of possible incompatibilities with Aztec, or take steps to minimize the effect of those changes.  In fact we already saw this being a problem since blocks were previously prefixed with `wp:core` (I believe), and now they're only prefixed with `wp:`.  Granted, the Gutenberg format may not be final, but that's the whole point of versioning if we ever need to change anything else.

All of the problems described above would easily be solved with explicit `wp_content` type info.

Reply:
Thanks for that @pento let us know if there's anything we can do to move it forward.

Reply:
Thanks for the details, @diegoreymendez. Some notes:

> The current check is not perfect and if a post written in wp-admin or Calypso describing Gutenberg included "<-- wp:" in its content, it would cause issues and probably data loss.

Actually, whatever the editor, if I want to write about "<-- wp:" in WordPress, the offending HTML entities would need to be escaped anyway. In other words, `post_content` will contain `&lt;-- wp`. Or did I miss something?

> In fact we already saw this being a problem since blocks were previously prefixed with `wp:core` (I believe), and now they're only prefixed with `wp:`.

The determining prefix has always been `<!-- wp:`, the full format being `<!-- wp:<namespace>/<block-name> <attributes>? -->`. There was indeed a change, but it was only adding a shorthand notation where `wp:foo` means `wp:core/foo`. I noticed in this PR that we used to be looking for `<!-- wp:core/`, but that would actually break non-core blocks even before the shorthand addition.

> Redundancy & synchronization, imperfect test condition, lack of proper versioning

I absolutely understand these concerns in abstract, but I'm not sure they are significant in this case. Here's why:
- I realize that there's a perception that the encoding schema for blocks hasn't been stable, but the grammar has actually been the same since December 2016. I don't see a need to synchronize the `contains( text, <!-- wp:` heuristic because it has been the same since Gutenberg became a plugin.
- Gutenberg represents a major shift forward for WordPress, and there has to be some concession for this to happen. Luckily (in my opinion), the project doesn't break away from the `post_content` model (as would be the case if we adopted data trees describing content as the source of truth, for instance), but the trade-off is that editors need to mindful of transformations performed on `post_content`.
- Some of these transformations have been ways of coping with certain quirks (`autop`) being one of those. With this in mind, the fact that the mobile implementation skips processing altogether if it detects the presence of blocks when opening a post doesn't tell me that we're bailing due to compatibility problems, but that we're heading towards trusting that `post_content` is good enough without the [mobile, in this case] editor tweaking it automatically.

Note that I'm not opposed to accommodating this in the REST API, but I'd like to weigh things a bit and provide context. :)

Reply:
> Actually, whatever the editor, if I want to write about "<-- wp:" in WordPress, the offending HTML entities would need to be escaped anyway. In other words, post_content will contain &lt;-- wp. Or did I miss something?

Just to clear this path of the discussion, I don't think escaping entities is an issue for us.  We should be good on that.

> There was indeed a change, but it was only adding a shorthand notation where wp:foo means wp:core/foo.

That caused issues in Aztec because we were looking for `<!-- wp:core` to identify Gutenberg blocks.  This is actually a good example of the kind of thing that could use some thinking because while you mention that `<!-- wp:` was the defining prefix, an external API consumer wouldn't know (unless this is very well documented somewhere).

> I realize that there's a perception that the encoding schema for blocks hasn't been stable, but the grammar has actually been the same since December 2016.

Is the grammar available through an API in a way that Aztec or other editors can use independently?

The thing that's important to remember is Aztec is an external editor that knows nothing about blocks or the Gutenberg grammar, and that only knows the standard HTML format.  We added some rules on our end to try and make it safe to use Gutenberg, but changes in the data stored in `wp_content` can break that quite easily.

> I don't see a need to synchronize the contains( text, <!-- wp: heuristic because it has been the same since Gutenberg became a plugin.

How will that work for external editors?  I'm can't really see how that being the same since the beginning is a good argument for not improving it.

> Gutenberg represents a major shift forward for WordPress, and there has to be some concession for this to happen.

I agree, but the concession here is really not necessary, if a change can make things better.

> Luckily (in my opinion), the project doesn't break away from the post_content model (as would be the case if we adopted data trees describing content as the source of truth, for instance), but the trade-off is that editors need to mindful of transformations performed on post_content.

Either we can always use the same rules to parse `wp_content` between Calypso, wp-admin and Gutenberg or we can't.  This PR is a very basic demonstration that some rules did change.

And the whole point of unifying the logic is to be mindful of changes we otherwise won't be aware of.

> Some of these transformations have been ways of coping with certain quirks (autop) being one of those. With this in mind, the fact that the mobile implementation skips processing altogether if it detects the presence of blocks when opening a post doesn't tell me that we're bailing due to compatibility problems, but that we're heading towards trusting that post_content is good enough without the [mobile, in this case] editor tweaking it automatically.

Not sure what you mean by tweaking the post, but:

- Processing a Gutenberg post through `autop` and `removep` breaks it.
- Processing a Calypso post without `autop` and `removep` breaks it.

How is that not a compatibility issue?

Reply:
@mcsf I'm not sure I understand what the problem is here. If we don't add a flag that marks posts as Gutenberg posts, and ideally indicates a version, we will have to add code to check for `<!-- wp:` at least on 4 different places:

- wp-admin (.com integration)
- Calypso (.com integration)
- iOS app
- Android app

I would prefer the flexibility that having a flag server side that all clients can use would give us, what do you think?

Reply:
The initial "Flags introduce synching issues" remark was directed at the idea there'd be some marker in the database indicating whether a post is block-based, which could become out-of-sync should the the `post_content` database field be modified directly.

([Related](https://make.wordpress.org/core/2017/01/17/editor-technical-overview/): "That’s another reason why `post_content` as source of truth matters, compared to storing a JSON structure in postmeta. Having things in two places means they can get out of sync depending on what tool you used to edit.")

A flag from an interface like the REST API is another thing, one which I don't have a particularly strong feeling for or against. On the surface, versioning seems a bit heavy-handed when in the strictest sense it's more about abandoning `autop` than anything about blocks or their syntax.

Reply:
Hey, @elibud and @diegoreymendez. I need to step out in a moment, but I'll be back explaining myself a bit better. In the meantime, I wanted to stress that I don't oppose at all to exposing a server-side check.

Reply:
What is not clear to me is how would we handle changes in `post_format` moving forward?

Our past examples are:
- `autop` and `removep` are not to be run on Gutenberg posts.
- `wp:core` being simplified to `wp:`.

If we had a versioning field we could at least just load the HTML editor for the incompatible post.

I realize this can feel a bit like overkill, but it would place some proper safeguards in place for format changes that we (and other potential external editors) could really use.

Reply:
PS: The synching issues would only happen if the editor modifying the content ignored the field - meaning that it's a bug on that editor.  While slightly more elaborated, it's a rather more robust approach than inspecting `post_content`.

Reply:
> What is not clear to me is how would we handle changes in `post_format` moving forward?
> 
> Our past examples are:
> 
> - `autop` and `removep` are not to be run on Gutenberg posts.
> - `wp:core` being simplified to `wp:`.
> If we had a versioning field we could at least just load the HTML editor for the incompatible post.
> 
> I realize this can feel a bit like overkill, but it would put some proper safeguards in place for format changes that we (and other potential external editors) could really use."

If there are format changes ahead, then we're doing something very wrong in Gutenberg. :-) The project is quickly approaching merge proposal and its language has been stable for a long time. There is simply no room (or need) to change the format of the thing that will be the basis for content production, persistence and display in all of WordPress. This is my way of stressing that there aren't going to be format changes[^1], not just because we don't want to, but because it would be totally unfeasible. Here's my take on the examples you cite:

- We debated the question of `autop` until later in the development of Gutenberg, and indeed changes to how we stored paragraphs in Gutenberg were made as late as last Fall. However, this has no impact in the block grammar, in the sense that a block is still defined by its `<!-- wp:` boundaries; the choice of whether to rely on `autop` only affects the content of a block. Put simply, the "only" thing that an editor that doesn't support Gutenberg must do is treat each block as a black box and not touch its inner HTML. Now, I'm using quotes because I know this is still a lot to ask and it's a major shift from how WordPress editors typically operate. But bear with me just a little longer.
- The simplification or aliasing of `wp:core/foo` to `wp:foo` should have had no impact on the way third parties recognize Gutenberg blocks. That is, looking for the `<!-- wp:core/` substring was the wrong choice in the first place, since that would limit the pattern to core blocks; it was a broken implementation even before the aliasing. Please don't take this statement the wrong way: I mean no judgment by it, and the mistake only highlights that the "language" of Gutenberg could have used more visibility. And this ties with what you've mentioned in a previous comment:

> while you mention that <!-- wp: was the defining prefix, an external API consumer wouldn't know (unless this is very well documented somewhere).

The good news is that this is something that's been discussed recently, and we're taking steps to make the grammar easier to read and to directly use in other systems. It's a PEG grammar, so any PEG parser can generate a parser for Gutenberg blocks, regardless of the environment (in Gutenberg, the same grammar generates a JS parser and a PHP parser). In practical terms, an editor like Aztec could use the parser like so (pseudocode):

```
// The grammar collects chunks that aren't wrapped with `<!-- wp` as "freeform"
// (in the UI, they correspond to the Classic block). The output of the parser
// is an array of blocks, where a block is an object with keys `blockName`,
// `attrs`, `innerBlocks`, `innerHTML`. Freeform blobs can be recognized as
// objects with no `blockName`.
process( text ) {
  return parseBlocks( text ).map( function( block ) {
    if ( ! block.blockName ) {
          // Only intervene in freeform blobs, leaving proper blocks untouched.
	  block.innerHTML = doAztecProcessing( block.innerHTML )
	}
	return block
  } )
}
```

> Is the grammar available through an API in a way that Aztec or other editors can use independently?

It's not available as an API, though I anticipate that there will be API endpoints to get a block-aware description of a post and related use cases. For an editor like Aztec, though, that sounds less relevant to me than the fact that the grammar is public (see my comment here https://github.com/wordpress-mobile/WordPress-iOS/pull/9194#discussion_r184233980 linking to the grammar proper and to a PR improving its visibility in the documentation).

> > I don't see a need to synchronize the `contains( text, <!-- wp:` heuristic because it has been the same since Gutenberg became a plugin.

> How will that work for external editors? I'm can't really see how that being the same since the beginning is a good argument for not improving it.

I wasn't making a case against improving, but pointing out that, since the heuristic hasn't changed and won't change, and since it's such a simple one — it's a substring check for a constant string — the question of _synchronization_ is, at least IMO, much less significant.

**To sum it all up :sweat_smile: :**

- For the purpose of simply preventing automatic reformatting of posts that contain Gutenberg blocks, the check for `<!-- wp:` is enough, *but* we can also work on a server-side check (though my personal opinion is that a remote test is a little overkill, since we've established that the test is trivially performed on the client and that the test string isn't going to change).

- If, hypothetically, we'd like Aztec to go further, it's absolutely possible to reuse the PEG-defined block grammar to parse content so as to avoid touching blocks and processing everything else.

- Post-Aztec, any editor interested in fully supporting Gutenberg can and should reuse said grammar.


[^1]: Individual blocks can, of course, change their internal formats. Gutenberg offers a migration path in its block API for this purpose. But those changes happen inside the black box of a block and thus would have no impact on a non-Gutenberg-ready editor.

Reply:
I don't think this discussion here is being constructive.  We have a very precise problem that can be solved with the hack we already have in this PR, or with proper versioning of a field for a robust technical solution.

Versioning means we'd have some level of support for automatically handling a lot of things in the future, from new post formats, to small changes in the current one.

This problem is **not theoretical** - we've already had a big backlash because of it in both iOS and Android.

In fact if there was no compatibility issue we wouldn't need this check to begin with.

Reply:
I suggest we switch to Slack if we need to discuss further, just to avoid flooding this PR.